REGRESSION(260774@main): [iPad] Video turns black after entering element fullscreen

jernoble opened the pull request:
#### e64627c9a4d48063a2c59af1f39fdb0195fc83cf
<pre>
REGRESSION(260774@main): [iPad] Video turns black after entering element fullscreen
<a href="https://bugs.webkit.org/show_bug.cgi?id=253225">https://bugs.webkit.org/show_bug.cgi?id=253225</a>
rdar://105990736

Reviewed by Tim Horton.

Prior to &quot;no-double-layer-hosting&quot;, the fullscreen machinery would create an AVPlayerLayerView and AVPlayerLayerViewController
simply to handle the case where the user might exit _element_ fullscreen when playing a video, and that video may need to
immediately enter picture-in-picture mode. As part of that process, the new AVPlayerLayerView would &quot;steal&quot; the video
content from an existing layer. This wasn&apos;t a problem in the double layer hosting world; there wasn&apos;t anything to steal then,
as the video layer being stolen was empty.

However, in the &quot;no-double-layer-hosting&quot; world, stealing the video layer results in no video being visible in the WKWebView,
as that video layer is the one inserted into the UI-side compositing tree. This code could benefit from a complete re-write,
now that there&apos;s a pre-existing AVPlayerLayerView in the web compositing hierarchy. For now, simply work around the problem
by not stealing the video layer prematurely, and by notifying the VideoFullscreenManagerProxy when PiP has ended and the
video layer can be returned to its rightful owner.

* Source/WebCore/platform/cocoa/VideoFullscreenChangeObserver.h:
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.mm:
(VideoFullscreenInterfaceAVKit::doSetup):
(VideoFullscreenInterfaceAVKit::returnToStandby):
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
(WebKit::VideoFullscreenModelContext::returnToStandby):
(WebKit::VideoFullscreenManagerProxy::returnToStandby):

Canonical link: <a href="https://commits.webkit.org/261143@main">https://commits.webkit.org/261143@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/216fb84463af6649c3738b74972d828fd473466c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110589 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19669 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43178 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1940 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119452 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114548 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21089 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10796 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/1437 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116331 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15709 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98898 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102853 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97669 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30566 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43979 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12320 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31904 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/86623 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12879 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8854 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18243 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51520 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7720 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14764 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
```diff
@@ -894,6 +900,23 @@ - (BOOL)prefersStatusBarHidden
     m_page->send(Messages::VideoFullscreenManager::ReturnVideoContentLayer(contextId));
 }
 
+void VideoFullscreenManagerProxy::returnToStandby(PlaybackSessionContextIdentifier contextId)
```
is "standby" a well-known term? it doesn't mean anything to me, and I wonder if a two or three word phrase might work better

jernoble replied:
I _wish_ I could name it something like "returnVideoContentLayer()", except that's already taken. :-/

jernoble replied:
And "standby" is a term of art for "PrepareForAutoPiPFromInline".

jernoble replied:
I'll rename this "returnVideoView()" so that it does what it says on the box.

jernoble committed:
REGRESSION(260774@main): [iPad] Video turns black after entering element fullscreen
https://bugs.webkit.org/show_bug.cgi?id=253225
rdar://105990736

Reviewed by Tim Horton.

Prior to "no-double-layer-hosting", the fullscreen machinery would create an AVPlayerLayerView and AVPlayerLayerViewController
simply to handle the case where the user might exit _element_ fullscreen when playing a video, and that video may need to
immediately enter picture-in-picture mode. As part of that process, the new AVPlayerLayerView would "steal" the video
content from an existing layer. This wasn't a problem in the double layer hosting world; there wasn't anything to steal then,
as the video layer being stolen was empty.

However, in the "no-double-layer-hosting" world, stealing the video layer results in no video being visible in the WKWebView,
as that video layer is the one inserted into the UI-side compositing tree. This code could benefit from a complete re-write,
now that there's a pre-existing AVPlayerLayerView in the web compositing hierarchy. For now, simply work around the problem
by not stealing the video layer prematurely, and by notifying the VideoFullscreenManagerProxy when PiP has ended and the
video layer can be returned to its rightful owner.

* Source/WebCore/platform/cocoa/VideoFullscreenChangeObserver.h:
* Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.mm:
(VideoFullscreenInterfaceAVKit::doSetup):
(VideoFullscreenInterfaceAVKit::returnToStandby):
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
(WebKit::VideoFullscreenModelContext::returnToStandby):
(WebKit::VideoFullscreenManagerProxy::returnToStandby):

Canonical link: https://commits.webkit.org/261143@main
